README.md
# php_utf8_converter
PHP Convert code to utf8


tools/ChangeLog
2006-07-31		First release for public version 1.0 

code/example/test.php
<?php
//require_once("config.php");
?>
<html>

<head>
<meta http-equiv="Content-Language" content="th">
<LINK href='../e-meeting/meeting_style.css' rel='stylesheet' type='text/css'>
</head>
<body class=link topmargin="3" leftmargin="3">
<?php
//ICONV("CP874"
echo iconv('MacThai', 'UTF-8', "§Ò¹½Ö¡ÍºÃÁ ½èÒÂºÃÔËÒÃ§Ò¹·ÑèÇä»");

?>
</body>
</html>

tools/utf8_db.php
<?
###########################################
## Phanupong Panyadee
## http://www.appservnetwork.com
## Convert Database from TIS-620 to UTF-8
## Powered by apples
##
## Version : 1.0
## Released : 2006-07-31
##
## HOW TO RUN THIS PROGRAM
##   1. This program design for "Command Line Execution" Only !!   ******** Not for WWW  ********.
##   2. This program will be "Convert all tables on your Database", Please specify database name below.
##   3. BEFORE Convert database please BACKUP YOUR DATABASE FIRST.
##   4. On Windows OS (AppServ).
##       - Goto Start -> Run -> cmd
##       - C:\AppServ\php\php.exe -d max_execution_time=0 utf8_db.php
##   5. On Linux (Test on Debian Linux).
##      - php -d max_execution_time=0 utf8_db.php
##
###########################################

$dbhost = "localhost";
$dbuname = "root";
$dbpass = "";
$dbname = "app";

$char1 = "TIS-620";	# Source Character Set 
$char2 = "UTF-8";		# Destination Character Set
$varchar_len = "0";	# 1 = Press y to confirm
										# 0 = Automatic Add Length for Varchar()

###########################################
# Use utf8_general_ci because it is faster.
# Use utf8_unicode_ci because it is more accurate.

$field_type = "utf8_unicode_ci"; 

###########################################
echo "<pre>";
if (!extension_loaded('iconv')) {
   if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
       dl('php_iconv.dll');
   } else {
       dl('iconv.so');
   }
}

mysql_connect($dbhost,$dbuname,$dbpass) or die ("Can't Connect to MySQL");
mysql_db_query($dbname,"ALTER DATABASE `$dbname` default character set utf8 collate $field_type");
$result=mysql_db_query($dbname,"show tables");
while(list($table)=mysql_fetch_row($result)) {
        $result2=mysql_db_query($dbname,"desc $table");
        echo "Table : $table\n";
        mysql_db_query($dbname,"ALTER TABLE `$table` DEFAULT CHARACTER SET utf8 COLLATE $field_type");
        while(list($field,$type,$null,$key)=mysql_fetch_row($result2)) {
			list($typename,$olen)=explode("(",$type);
			switch ($typename) {
				case "varchar" :
					$result3=mysql_db_query($dbname,"SELECT LENGTH(MAX( $field )) FROM  `$table`");
					list($max_len)=mysql_fetch_row($result3);
					$olen=substr($olen,0,-1);
					$len = $olen + 80;
					if ($varchar_len == "1") {
						echo "\nField : $field\n";
						echo "Varchar($olen) Need to change to Varchar($len) ?\n";
						echo "Type 'y' to confirm : ";
						$input = trim(strtolower(fgets(STDIN)));
					} else {
						$input = "y";
					}
					if ($input == "y") {
						mysql_db_query($dbname,"ALTER TABLE `$table` CHANGE `$field` `$field` VARCHAR( $len ) ");
					}
					unset ($olen,$len);
				break;
			}
                mysql_db_query($dbname,"ALTER TABLE `$table` CHANGE `$field` `$field` $type CHARACTER SET utf8 COLLATE $field_type");
                $all_field[] = $field;
                if ($key == "PRI") {
                        $table_key[] = $field;
                }
        }
        foreach ($all_field as $value) {
                $select_field .= "`$value`,"; 
        }
        $num_field=count($all_field);
        $num_key=count($table_key);
        $select_field = substr($select_field ,0,-1);
        $update_field = substr($update_field ,0,-18);
        mysql_db_query($dbname,"SET NAMES latin1");
        mysql_db_query($dbname,"SET collation_connection = 'latin1_swedish_ci'");
        $result3=mysql_db_query($dbname,"select $select_field from $table");
        mysql_db_query($dbname,"SET NAMES utf8");
        mysql_db_query($dbname,"SET collation_connection = '$field_type'");
        $arow=mysql_num_rows($result3);
        $i = 1;
        if ($arow !=0) {
                while($arr=mysql_fetch_array($result3)){
                        if ($i == 1) {
							echo "   Found $arow fields.\n";
                        }
                        ++$i;
                        for ($x=0;$x<$num_field;$x++) {
							if (is_null($arr[$x])) {
								$var = "`$all_field[$x]`=NULL,";
							} else {
                                $var = ICONV("$char1","$char2",$arr[$x]);
								$lenchar1=strlen($arr[$x]);
								$lenchar2=strlen_utf8($var);
								if ($char1 == "TIS-620") {	 // Loop for Thai Convert. You can split to function.
									if ($lenchar1 != $lenchar2) {
										++$no_len;
										# echo "$no_len Char Len : $lenchar1 && $lenchar2 ";
										$var = ICONV("CP874","$char2",$arr[$x]);
										$lenchar2=strlen_utf8($var);
										if ($lenchar1 == $lenchar2) {
											$result_len = "Ok";
										} else {
											$result_len = "MacThai";
										}
										if ($result_len == "MacThai") {
											$var = ICONV("MacThai","$char2",$arr[$x]);
											$lenchar2=strlen_utf8($var);
											if ($lenchar1 == $lenchar2) {
												$result_len = "Ok";
											} else {
												$result_len = "Another Charset";
												$var = $arr[$x];
											}
										}
										# echo "Again : $lenchar2 $result_len\n";
									} // end if $lenchar1 != $lenchar2
                                }
                                $var = ereg_replace("\\\\","\\\\",$var);
                                $var = ereg_replace("'","\'",$var);
                                $var = "`$all_field[$x]`='$var',";
							}
                                $update_field .= $var;
                                if (isset($table_key)) {
									if ($x==0) {
										for ($y=0;$y<$num_key;$y++) {
											$key = $table_key[$y];
											$update_where .= "`$table_key[$y]`='$arr[$key]' and ";
										}
									}
                                } else {
									$update_where .= "`$all_field[$x]`='$arr[$x]' and ";
                                }
                        }
                        $update_field = substr($update_field ,0,-1);
                        $update_where = substr($update_where ,0,-4);
                        #echo "UPDATE $table  SET $update_field where $update_where\n";
                        $uresult=mysql_db_query($dbname,"UPDATE $table  SET $update_field where $update_where");
						$converted = $converted + $uresult;
                        unset($update_where,$update_field);
                }
        } else {
                echo "   NO Data in this table. \n";
        }
         echo "   Converted $converted fields.\n";
        echo "------------------------\n";
        unset($all_field,$select_field,$update_field,$table_key,$converted);
}

function strlen_utf8 ($str) {
   $i = 0;
   $count = 0;
   $len = strlen ($str);
   while ($i < $len) {
		$chr = ord ($str[$i]);
		$count++;
		$i++;
		if ($i >= $len)   break;
			if ($chr & 0x80) {
			   $chr <<= 1;
			   while ($chr & 0x80) {
					$i++;
					$chr <<= 1;
			   }
			}
		}
   return $count;
}
?>

application/views/index.php
<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3" crossorigin="anonymous">
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.6.0/jquery.min.js"></script>

<style>
label {
	font-weight: bold;
	margin-bottom: 3px;
}
</style>
<div class="container">
<?php
if(!$target_path) $target_path = "C:\\xampp3\\htdocs\\encoder\\code\\example\\";

echo "<h2>Target path:</b> ".$target_path."</h2>";

$dir = directory_map($target_path);
?>
<br>
<form method="post">
<p>
	<label>Target path</label>
	<input type="text" name="target" value="<?php echo $target_path; ?>" class="form-control">
</p>
<p>
	<button type="button" name="get" class="btn btn-primary">Get File</button>
	<button type="submit" name="convert" class="btn btn-primary">Convert</button>
</p>
</form>

<?php
echo "<h2>Item Lists</h2>";

if($dir)
foreach($dir as $ind => $val) {
	if(is_array($val)) {
		echo "<b>".$ind."</b><br>";		
	} else {
		//echo $val."<br>";
	}
}

foreach($dir as $ind => $val) {
	if(is_array($val)) {
		//echo "<b>".$ind."</b><br>";		
	} else {
		echo $val."<br>";
	}
}
?>
</div>

<script>
$(document).ready(function() {
	$("button[name=get]").click(function() {
		window.location = "?path=" + $("input[name=target]").val();
	});
});
</script>

application/views/convert-file.php
<?php
###########################################
## Phanupong Panyadee
## http://www.appservnetwork.com
## Convert File from TIS-620 to UTF-8
## Powered by apples
##
## Version : 1.0
## Released : 2006-07-31
##
## HOW TO RUN THIS PROGRAM
##   1. This program design for WWW.
##   2. Access via www e.g. http://www.appservnetwork.com/free-scripts/utf-8/utf8_convert_file.php
##
###########################################

$char1 = "TIS-620";		# Source Character Set TIS-620 Please Leave BLANK !!
$char2 = "UTF-8";			# Destination Character Set
$output = "";
###########################################
if (!extension_loaded('iconv')) {
   if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
       dl('php_iconv.dll');
   } else {
       dl('iconv.so');
   }
}

echo $path."<br>";
return false;

if($_FILES) {
foreach($_FILES as $key => $value) {
	$file_name = $value["name"];
	$file_size = $value["size"];
	$file_type = $value["type"];
	$file_tmp_name = $value["tmp_name"];
}

$numfile=count($file_name);
if ($numfile != 0) {
	for ($i=0;$i<$numfile;$i++) {
		$arr=file($file_tmp_name[$i]);
		$count=count($arr);
		for ($x=0;$x<$count;$x++) {
			$var = ICONV("$char1","$char2",$arr[$x]);
			$lenchar1=strlen($arr[$x]);
			$lenchar2=strlen_utf8($var);
			if ($char1 == "TIS-620") {	 // Loop for Thai Convert. You can split to function.
				if ($lenchar1 != $lenchar2) {
					++$no_len;
					$var = ICONV("CP874","$char2",$arr[$x]);
					$lenchar2=strlen_utf8($var);
					if ($lenchar1 == $lenchar2) {
						$result_len = "Ok";
					} else {
						$result_len = "MacThai";
					}
					if ($result_len == "MacThai") {
						$var = ICONV("MacThai","$char2",$arr[$x]);
						$lenchar2=strlen_utf8($var);
						if ($lenchar1 == $lenchar2) {
							$result_len = "Ok";
						} else {
							$result_len = "Another Charset";
							$var = $arr[$x];
						}
					}
				} // end if $lenchar1 != $lenchar2
			}
			$output .= $var;
		}
		$handle = fopen($file_tmp_name[$i],"w");
		fwrite($handle,$output);
		fclose($handle);
		header("Pragma: public");
		header("Expires: 0");
		header("Cache-Control: must-revalidate, post-check=0, pre-check=0"); 
		header("Content-Type: application/force-download");
		header("Content-Type: application/octet-stream");
		header("Content-Type: application/download");
		header("Content-Disposition: attachment; filename=".basename("$file_name[$i]").";");
		header("Content-Transfer-Encoding: binary\r\n");
		header("Content-Length: ".filesize($file_tmp_name[$i]));
		readfile($file_tmp_name[$i]);
		unlink($file_tmp_name[$i]);
	}
	exit();
}
}

function strlen_utf8 ($str) {
   $i = 0;
   $count = 0;
   $len = strlen ($str);
   while ($i < $len) {
		$chr = ord ($str[$i]);
		$count++;
		$i++;
		if ($i >= $len)   break;
			if ($chr & 0x80) {
			   $chr <<= 1;
			   while ($chr & 0x80) {
					$i++;
					$chr <<= 1;
			   }
			}
		}
   return $count;
}
?>

application/controllers/Welcome.php
<?php
defined('BASEPATH') OR exit('No direct script access allowed');

class Welcome extends CI_Controller {

	/**
	 * Index Page for this controller.
	 *
	 * Maps to the following URL
	 * 		http://example.com/index.php/welcome
	 *	- or -
	 * 		http://example.com/index.php/welcome/index
	 *	- or -
	 * Since this controller is set as the default controller in
	 * config/routes.php, it's displayed at http://example.com/
	 *
	 * So any other public methods not prefixed with an underscore will
	 * map to /index.php/welcome/<method_name>
	 * @see https://codeigniter.com/userguide3/general/urls.html
	 */
	public function __construct()
	{
		parent::__construct();
		// Your own constructor code
		$this->load->helper("file");
		$this->load->helper('directory');
	}

	public function index()
	{
		$convert = $this->input->post("convert");
		$target = $this->input->post("target");

		if($target) {
			if(!is_dir($target."/_output")) {
				mkdir($target."/_output");
			}

			//$dir = directory_map($target_path);
			$this->clone_dir($target."/_output/", $target);

			$this->convert($target."/_output/", $target);

			$this->load->view('convert');
		} else {
			$this->load->view('index');
		}
	}

	public function clone_dir($dist="_output", $target_path)
	{
		$dir = directory_map($target_path);

		if($dir)
		foreach($dir as $ind => $val) {
			if(is_array($val)) {
				if($ind != '_output\\') {
					@mkdir($dist."/".$ind); 
					echo "Create Folder :: ".$dist."/".$ind."<br>";
					$this->clone_dir($dist."/".$ind, $target_path."/".$ind);
				}
			} else {

			}
		}
	}

	public function convert($dist="_output", $target_path)
	{
		$dir = directory_map($target_path);

		if($dir)
		foreach($dir as $ind => $val) {
			if(is_array($val)) {
				if($ind != '_output\\') {
					//mkdir($dist."/".$ind); 
					//echo "Create :: ".$dist."/".$ind."<br>";
					$this->convert($dist."/".$ind, $target_path."/".$ind);
				}
			} else {
				$file_path = $target_path."/".$val;
				$this->utf8_convert($file_path, $dist);

				//$this->load->view("convert-file", $data);
			}
		}
	}

	public function utf8_convert($file_path, $dest="")
	{
		if (!extension_loaded('iconv')) {
		   if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
			   dl('php_iconv.dll');
		   } else {
			   dl('iconv.so');
		   }
		}

		echo "utf8_convert :: ".$file_path."<br>";

		$char1 = "TIS-620";		# Source Character Set TIS-620 Please Leave BLANK !!
		$char1 = "MacThai";		# Source Character Set TIS-620 Please Leave BLANK !!
		
		$char2 = "UTF-8";			# Destination Character Set
		$output = "";

		$file_info = get_file_info($file_path);
		echo $file_path."<br>";
		//var_dump($file_info);
		
		$file_name = $file_info["name"];
		$file_size = $file_info["size"];
		//$file_type = $file_info["type"];
		//$file_tmp_name = $file_info["tmp_name"];

		$arr=file($file_path);
		$count=count($arr);
		for ($x=0;$x<$count;$x++) {
			//echo $char1."=>".$char2."<br>";
			$var = ICONV($char1, $char2, $arr[$x]);
			$lenchar1=strlen($arr[$x]);
			$lenchar2=$this->strlen_utf8($var);
			if ($char1 == "TIS-620") {	 // Loop for Thai Convert. You can split to function.
				if ($lenchar1 != $lenchar2) {
					++$no_len;
					$var = ICONV("CP874", $char2, $arr[$x]);
					$lenchar2=$this->strlen_utf8($var);
					if ($lenchar1 == $lenchar2) {
						$result_len = "Ok";
					} else {
						$result_len = "MacThai";
					}
					if ($result_len == "MacThai") {
						$var = ICONV("MacThai", $char2 ,$arr[$x]);
						$lenchar2=$this->strlen_utf8($var);
						if ($lenchar1 == $lenchar2) {
							$result_len = "Ok";
						} else {
							$result_len = "Another Charset";
							$var = $arr[$x];
						}
					}
				} // end if $lenchar1 != $lenchar2
			}
			$output .= $var;
		}
		
		$handle = fopen($dest."/".$file_name,"w");
		fwrite($handle, $output);
		fclose($handle);

		echo "destination :: ".$dest."/".$file_name."<br>";
	}

	public function strlen_utf8($str) 
	{
	   $i = 0;
	   $count = 0;
	   $len = strlen ($str);
	   while ($i < $len) {
			$chr = ord ($str[$i]);
			$count++;
			$i++;
			if ($i >= $len)   break;
				if ($chr & 0x80) {
				   $chr <<= 1;
				   while ($chr & 0x80) {
						$i++;
						$chr <<= 1;
				   }
				}
			}
	   return $count;
	}
}


tools/utf8_convert_file.php
<?php
###########################################
## Phanupong Panyadee
## http://www.appservnetwork.com
## Convert File from TIS-620 to UTF-8
## Powered by apples
##
## Version : 1.0
## Released : 2006-07-31
##
## HOW TO RUN THIS PROGRAM
##   1. This program design for WWW.
##   2. Access via www e.g. http://www.appservnetwork.com/free-scripts/utf-8/utf8_convert_file.php
##
###########################################

$char1 = "TIS-620";		# Source Character Set TIS-620 Please Leave BLANK !!
$char2 = "UTF-8";			# Destination Character Set
$output = "";
###########################################
if (!extension_loaded('iconv')) {
   if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
       dl('php_iconv.dll');
   } else {
       dl('iconv.so');
   }
}
if($_FILES) {
foreach($_FILES as $key => $value) {
	$file_name = $value["name"];
	$file_size = $value["size"];
	$file_type = $value["type"];
	$file_tmp_name = $value["tmp_name"];
}

$numfile=count($file_name);
if ($numfile != 0) {
	for ($i=0;$i<$numfile;$i++) {
		$arr=file($file_tmp_name[$i]);
		$count=count($arr);
		for ($x=0;$x<$count;$x++) {
			$var = ICONV("$char1","$char2",$arr[$x]);
			$lenchar1=strlen($arr[$x]);
			$lenchar2=strlen_utf8($var);
			if ($char1 == "TIS-620") {	 // Loop for Thai Convert. You can split to function.
				if ($lenchar1 != $lenchar2) {
					++$no_len;
					$var = ICONV("CP874","$char2",$arr[$x]);
					$lenchar2=strlen_utf8($var);
					if ($lenchar1 == $lenchar2) {
						$result_len = "Ok";
					} else {
						$result_len = "MacThai";
					}
					if ($result_len == "MacThai") {
						$var = ICONV("MacThai","$char2",$arr[$x]);
						$lenchar2=strlen_utf8($var);
						if ($lenchar1 == $lenchar2) {
							$result_len = "Ok";
						} else {
							$result_len = "Another Charset";
							$var = $arr[$x];
						}
					}
				} // end if $lenchar1 != $lenchar2
			}
			$output .= $var;
		}
		$handle = fopen($file_tmp_name[$i],"w");
		fwrite($handle,$output);
		fclose($handle);
		header("Pragma: public");
		header("Expires: 0");
		header("Cache-Control: must-revalidate, post-check=0, pre-check=0"); 
		header("Content-Type: application/force-download");
		header("Content-Type: application/octet-stream");
		header("Content-Type: application/download");
		header("Content-Disposition: attachment; filename=".basename("$file_name[$i]").";");
		header("Content-Transfer-Encoding: binary\r\n");
		header("Content-Length: ".filesize($file_tmp_name[$i]));
		readfile($file_tmp_name[$i]);
		unlink($file_tmp_name[$i]);
	}
	exit();
}
}

echo "<html>";
echo "<head>";
echo "<title>Convert file to UTF-8 format Powered by apples</title>";
echo "<META HTTP-EQUIV=\"Content-Type\" CONTENT=\"text/html; charset=UTF-8\">";
echo "</head>";
echo "<body>";
echo "<form action=\"utf8_convert_file.php\" method=\"post\" enctype=\"multipart/form-data\">";
echo "<table border=\"0\" width=\"332\" cellspacing=\"1\" bgcolor=\"#E6F2FF\" align=\"center\">";
echo "<tr><td width=\"332\" height=\"30\" align=\"center\" bgcolor=\"#DFEFFF\"><strong>Convert file to UTF-8 format.</strong></td></tr>";
echo "<tr><td width=\"332\" valign=\"top\" align=\"center\" bgcolor=\"#F4FAFF\">";
echo "<table border=\"0\" width=\"296\" cellspacing=\"1\">";
echo "<tr><td width=\"81\" height=\"26\">File 1 : </td>";
echo "<td width=\"203\" height=\"26\"><input type=\"file\" name=\"file[]\"><br></td>";
echo "</tr>";
echo "</table>";
echo "</td>";
echo "</tr>";
echo "<tr><td width=\"332\" height=\"35\" align=\"center\" bgcolor=\"#F9FCFF\"><input type=\"submit\" value=\"Convert Me !\"></td>";
echo "</tr>";
echo "</table>";
echo "</form>";
echo "<br><br><center>Powered by apples <br>";
echo "<a href=\"http://www.appservnetwork.com\">http://www.appservnetwork.com</a></html>";

function strlen_utf8 ($str) {
   $i = 0;
   $count = 0;
   $len = strlen ($str);
   while ($i < $len) {
		$chr = ord ($str[$i]);
		$count++;
		$i++;
		if ($i >= $len)   break;
			if ($chr & 0x80) {
			   $chr <<= 1;
			   while ($chr & 0x80) {
					$i++;
					$chr <<= 1;
			   }
			}
		}
   return $count;
}
?>